{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f691c18-e866-4699-ac62-faf1f0775d00",
   "metadata": {},
   "outputs": [],
   "source": [
    "Description of Template\n",
    "This template is designed to help you get started with the All by All tables. It streamlines data import and provides a quick overview of the data structure.\n",
    "\n",
    "Load Libraries\n",
    "import hail as hl\n",
    "import os\n",
    "Loading BokehJS ...\n",
    "Hail Matrix Table Import\n",
    "The Hail Matrix Tables contain the locus for each gene or variant. You can use these to explore all possible phenotypes associated with the gene or variant.\n",
    "\n",
    "POP specifies either single ancestry results or meta-analysis results.\n",
    "\n",
    "The following values can be assigned: AFR, AMR, EAS, EUR, MID, SAS, META.\n",
    "TYPE specifies the type of association test.\n",
    "\n",
    "The following values can be assigned: ACAF, exome, gene.\n",
    "############ EDIT POP AND TYPE VALUES ############\n",
    "\n",
    "POP = \"SAS\" # south asian ancestry\n",
    "TYPE = \"ACAF\" # greater than 1% population-specific allele freq\n",
    "############ DO NOT EDIT THIS CODE ############\n",
    "\n",
    "# Import GWAS results for ACAF variants for participants with SAS genetic ancestry:\n",
    "SAS_ACAF_mt = hl.read_matrix_table(f\"gs://fc-aou-datasets-controlled/AllxAll/v1/mt/{POP}_{TYPE}_results.mt\")\n",
    "Initializing Hail with default parameters...\n",
    "/opt/conda/lib/python3.10/site-packages/hailtop/aiocloud/aiogoogle/user_config.py:43: UserWarning:\n",
    "\n",
    "Reading spark-defaults.conf to determine GCS requester pays configuration. This is deprecated. Please use `hailctl config set gcs_requester_pays/project` and `hailctl config set gcs_requester_pays/buckets`.\n",
    "\n",
    "Setting default log level to \"WARN\".\n",
    "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n",
    "Running on Apache Spark version 3.3.0\n",
    "SparkUI available at http://all-of-us-6701-m.us-central1-c.c.terra-vpc-sc-2cfda53d.internal:42333\n",
    "Welcome to\n",
    "     __  __     <>__\n",
    "    / /_/ /__  __/ /\n",
    "   / __  / _ `/ / /\n",
    "  /_/ /_/\\_,_/_/_/   version 0.2.130.post1-c69cd67afb8b\n",
    "LOGGING: writing to /home/jupyter/workspaces/duplicateofhslscodetemplates/hail-20241016-1834-0.2.130.post1-c69cd67afb8b.log\n",
    "You can look at the structure of the Hail Matrix Table using the .describe() function.\n",
    "\n",
    "SAS_ACAF_mt.describe()\n",
    "----------------------------------------\n",
    "Global fields:\n",
    "    None\n",
    "----------------------------------------\n",
    "Column fields:\n",
    "    'phenoname': str\n",
    "    'n_cases': int64\n",
    "    'n_controls': int64\n",
    "    'heritability': float64\n",
    "    'saige_version': str\n",
    "    'inv_normalized': str\n",
    "    'pheno_sex': str\n",
    "    'trait_type': str\n",
    "    'category': str\n",
    "    'pop': str\n",
    "    'description': str\n",
    "    'phecode_category': str\n",
    "    'description_more': str\n",
    "    'lambda_gc': float64\n",
    "----------------------------------------\n",
    "Row fields:\n",
    "    'locus': locus<GRCh38>\n",
    "    'alleles': array<str>\n",
    "    'MarkerID': str\n",
    "    'annotation': str\n",
    "    'variant_id': str\n",
    "    'gene_id': str\n",
    "    'gene_symbol': str\n",
    "    'transcript_id': str\n",
    "    'vep': struct {\n",
    "        allele_string: str, \n",
    "        colocated_variants: array<struct {\n",
    "            allele_string: str, \n",
    "            clin_sig: array<str>, \n",
    "            clin_sig_allele: str, \n",
    "            end: int32, \n",
    "            id: str, \n",
    "            phenotype_or_disease: int32, \n",
    "            pubmed: array<int32>, \n",
    "            somatic: int32, \n",
    "            start: int32, \n",
    "            strand: int32\n",
    "        }>, \n",
    "        end: int32, \n",
    "        id: str, \n",
    "        input: str, \n",
    "        intergenic_consequences: array<struct {\n",
    "            allele_num: int32, \n",
    "            consequence_terms: array<str>, \n",
    "            impact: str, \n",
    "            variant_allele: str\n",
    "        }>, \n",
    "        most_severe_consequence: str, \n",
    "        motif_feature_consequences: array<struct {\n",
    "            allele_num: int32, \n",
    "            consequence_terms: array<str>, \n",
    "            high_inf_pos: str, \n",
    "            impact: str, \n",
    "            motif_feature_id: str, \n",
    "            motif_name: str, \n",
    "            motif_pos: int32, \n",
    "            motif_score_change: float64, \n",
    "            transcription_factors: array<str>, \n",
    "            strand: int32, \n",
    "            variant_allele: str\n",
    "        }>, \n",
    "        regulatory_feature_consequences: array<struct {\n",
    "            allele_num: int32, \n",
    "            biotype: str, \n",
    "            consequence_terms: array<str>, \n",
    "            impact: str, \n",
    "            regulatory_feature_id: str, \n",
    "            variant_allele: str\n",
    "        }>, \n",
    "        seq_region_name: str, \n",
    "        start: int32, \n",
    "        strand: int32, \n",
    "        transcript_consequences: array<struct {\n",
    "            allele_num: int32, \n",
    "            amino_acids: str, \n",
    "            appris: str, \n",
    "            biotype: str, \n",
    "            canonical: int32, \n",
    "            ccds: str, \n",
    "            cdna_start: int32, \n",
    "            cdna_end: int32, \n",
    "            cds_end: int32, \n",
    "            cds_start: int32, \n",
    "            codons: str, \n",
    "            consequence_terms: array<str>, \n",
    "            distance: int32, \n",
    "            domains: array<struct {\n",
    "                db: str, \n",
    "                name: str\n",
    "            }>, \n",
    "            exon: str, \n",
    "            flags: str, \n",
    "            gene_id: str, \n",
    "            gene_pheno: int32, \n",
    "            gene_symbol: str, \n",
    "            gene_symbol_source: str, \n",
    "            hgnc_id: str, \n",
    "            hgvsc: str, \n",
    "            hgvsp: str, \n",
    "            hgvs_offset: int32, \n",
    "            impact: str, \n",
    "            intron: str, \n",
    "            lof: str, \n",
    "            lof_flags: str, \n",
    "            lof_filter: str, \n",
    "            lof_info: str, \n",
    "            mane_select: str, \n",
    "            mane_plus_clinical: str, \n",
    "            mirna: array<str>, \n",
    "            polyphen_prediction: str, \n",
    "            polyphen_score: float64, \n",
    "            protein_end: int32, \n",
    "            protein_start: int32, \n",
    "            protein_id: str, \n",
    "            sift_prediction: str, \n",
    "            sift_score: float64, \n",
    "            source: str, \n",
    "            strand: int32, \n",
    "            transcript_id: str, \n",
    "            tsl: int32, \n",
    "            uniprot_isoform: array<str>, \n",
    "            variant_allele: str\n",
    "        }>, \n",
    "        variant_class: str\n",
    "    }\n",
    "    'worst_csq_by_gene_canonical': array<struct {\n",
    "        allele_num: int32, \n",
    "        amino_acids: str, \n",
    "        appris: str, \n",
    "        biotype: str, \n",
    "        canonical: int32, \n",
    "        ccds: str, \n",
    "        cdna_start: int32, \n",
    "        cdna_end: int32, \n",
    "        cds_end: int32, \n",
    "        cds_start: int32, \n",
    "        codons: str, \n",
    "        consequence_terms: array<str>, \n",
    "        distance: int32, \n",
    "        domains: array<struct {\n",
    "            db: str, \n",
    "            name: str\n",
    "        }>, \n",
    "        exon: str, \n",
    "        flags: str, \n",
    "        gene_id: str, \n",
    "        gene_pheno: int32, \n",
    "        gene_symbol: str, \n",
    "        gene_symbol_source: str, \n",
    "        hgnc_id: str, \n",
    "        hgvsc: str, \n",
    "        hgvsp: str, \n",
    "        hgvs_offset: int32, \n",
    "        impact: str, \n",
    "        intron: str, \n",
    "        lof: str, \n",
    "        lof_flags: str, \n",
    "        lof_filter: str, \n",
    "        lof_info: str, \n",
    "        mane_select: str, \n",
    "        mane_plus_clinical: str, \n",
    "        mirna: array<str>, \n",
    "        polyphen_prediction: str, \n",
    "        polyphen_score: float64, \n",
    "        protein_end: int32, \n",
    "        protein_start: int32, \n",
    "        protein_id: str, \n",
    "        sift_prediction: str, \n",
    "        sift_score: float64, \n",
    "        source: str, \n",
    "        strand: int32, \n",
    "        transcript_id: str, \n",
    "        tsl: int32, \n",
    "        uniprot_isoform: array<str>, \n",
    "        variant_allele: str, \n",
    "        most_severe_consequence: str, \n",
    "        csq_score: float64\n",
    "    }>\n",
    "    'revel': float64\n",
    "    'splice_ai_acceptor_gain_score': float64\n",
    "    'splice_ai_acceptor_gain_distance': int32\n",
    "    'splice_ai_acceptor_loss_score': float64\n",
    "    'splice_ai_acceptor_loss_distance': int32\n",
    "    'splice_ai_donor_gain_score': float64\n",
    "    'splice_ai_donor_gain_distance': int32\n",
    "    'splice_ai_donor_loss_score': float64\n",
    "    'splice_ai_donor_loss_distance': int32\n",
    "    'hgvsp': str\n",
    "    'splice_ai_ds': float64\n",
    "    'AF_raw': array<float64>\n",
    "    'AC_raw': array<int32>\n",
    "    'AN_raw': int32\n",
    "    'homozygote_count_raw': array<int32>\n",
    "    'AF': array<float64>\n",
    "    'AC': array<int32>\n",
    "    'AN': int32\n",
    "    'homozygote_count': array<int32>\n",
    "    'quality_flags': struct {\n",
    "        hq_exp_AC_variant: bool\n",
    "    }\n",
    "    'quality_flags_lambda': struct {\n",
    "        hq_AF_variant: bool\n",
    "    }\n",
    "    'hq_variant': bool\n",
    "    'hq_variant_lambda': bool\n",
    "----------------------------------------\n",
    "Entry fields:\n",
    "    'AC_Allele2': int32\n",
    "    'AF_Allele2': float64\n",
    "    'MissingRate': float64\n",
    "    'BETA': float64\n",
    "    'SE': float64\n",
    "    'var': float64\n",
    "    'p.value.NA': float64\n",
    "    'Is.SPA': bool\n",
    "    'AF_case': float64\n",
    "    'AF_ctrl': float64\n",
    "    'Pvalue': float64\n",
    "    'hq_exp_AC': bool\n",
    "----------------------------------------\n",
    "Column key: ['phenoname']\n",
    "Row key: ['locus', 'alleles']\n",
    "----------------------------------------\n",
    "Hail Table Import\n",
    "The Hail Tables contain the GWAS results for a single phenotype. You can use these to explore all possible variants or genes associated with a certain phenotype.\n",
    "\n",
    "POP specifies either single ancestry results or meta-analysis results.\n",
    "\n",
    "The following values can be assigned: AFR, AMR, EAS, EUR, MID, SAS, META.\n",
    "TYPE specifies the type of association test.\n",
    "\n",
    "The following values can be assigned: ACAF, exome, gene.\n",
    "PHENONAME specifies the phenotype of interest by concept ID.\n",
    "\n",
    "There are many concept IDs for phenotypes. Please see the References section to find the concept ID for your phenotype of interest.\n",
    "############ EDIT POP, TYPE, AND PHENONAME VALUES ############\n",
    "\n",
    "POP = \"EUR\" # south asian ancestry\n",
    "TYPE = \"ACAF\" # greater than 1% population-specific allele freq\n",
    "PHENONAME = \"1740608\" # acne phenotype code\n",
    "############ DO NOT EDIT THIS CODE ############\n",
    "\n",
    "# The results table used for the example will be the ACAF & European results for the phenotype \"acne\":\n",
    "EUR_ACAF_acne_ht = hl.read_table(f\"gs://fc-aou-datasets-controlled/AllxAll/v1/ht/{TYPE}/{POP}/phenotype_{PHENONAME}_{TYPE}_results.ht\")\n",
    "Since Hail Tables are in a tabular format, you can use the function .show(5) to display the first 5 rows of the data.\n",
    "\n",
    "EUR_ACAF_acne_ht.show(5)\n",
    "[Stage 0:>                                                          (0 + 4) / 4]\n",
    "CHR\n",
    "POS\n",
    "MarkerID\n",
    "Allele1\n",
    "Allele2\n",
    "AC_Allele2\n",
    "AF_Allele2\n",
    "MissingRate\n",
    "BETA\n",
    "SE\n",
    "var\n",
    "p.value.NA\n",
    "Is.SPA\n",
    "AF_case\n",
    "AF_ctrl\n",
    "locus\n",
    "alleles\n",
    "phenoname\n",
    "Pvalue\n",
    "Pvalue_log10\n",
    "rank\n",
    "Pvalue_expected\n",
    "Pvalue_expected_log10\n",
    "str\tint32\tstr\tstr\tstr\tint32\tfloat64\tfloat64\tfloat64\tfloat64\tfloat64\tfloat64\tbool\tfloat64\tfloat64\tlocus<GRCh38>\tarray<str>\tstr\tfloat64\tfloat64\tint64\tfloat64\tfloat64\n",
    "\"chr1\"\t10447\t\"chr1:10447_CCCTAA/C\"\t\"CCCTAA\"\t\"C\"\t20\t1.43e-04\t9.71e-02\t7.13e-01\t1.33e+00\t5.70e-01\t5.91e-01\tFalse\t2.52e-04\t1.40e-04\tchr1:10447\t[\"CCCTAA\",\"C\"]\t\"1740608\"\t5.91e-01\t2.29e-01\t21129587\t4.61e-01\t3.36e-01\n",
    "\"chr1\"\t13668\t\"chr1:13668_G/A\"\t\"G\"\t\"A\"\t34\t2.43e-04\t1.45e-01\t-5.20e-01\t7.22e-01\t1.92e+00\t4.72e-01\tFalse\t0.00e+00\t2.50e-04\tchr1:13668\t[\"G\",\"A\"]\t\"1740608\"\t4.72e-01\t3.26e-01\t16025452\t3.50e-01\t4.56e-01\n",
    "\"chr1\"\t14487\t\"chr1:14487_G/A\"\t\"G\"\t\"A\"\t2\t1.43e-05\t4.47e-02\t-1.03e+00\t4.93e+00\t4.12e-02\t8.35e-01\tFalse\t0.00e+00\t1.47e-05\tchr1:14487\t[\"G\",\"A\"]\t\"1740608\"\t8.35e-01\t7.82e-02\t36932725\t8.06e-01\t9.35e-02\n",
    "\"chr1\"\t14671\t\"chr1:14671_G/C\"\t\"G\"\t\"C\"\t294\t2.10e-03\t8.10e-02\t-7.84e-01\t3.57e-01\t7.73e+00\t2.93e-02\tTrue\t5.04e-04\t2.15e-03\tchr1:14671\t[\"G\",\"C\"]\t\"1740608\"\t2.81e-02\t1.55e+00\t1155029\t2.52e-02\t1.60e+00\n",
    "\"chr1\"\t16226\t\"chr1:16226_AG/A\"\t\"AG\"\t\"A\"\t96\t6.87e-04\t1.49e-02\t8.60e-02\t6.13e-01\t2.66e+00\t8.88e-01\tFalse\t7.56e-04\t6.85e-04\tchr1:16226\t[\"AG\",\"A\"]\t\"1740608\"\t8.88e-01\t5.14e-02\t41673130\t9.10e-01\t4.11e-02\n",
    "showing top 5 rows\n",
    "\n",
    "To look at the structure of the Hail Table, you can use the .describe() function.\n",
    "\n",
    "EUR_ACAF_acne_ht.describe()\n",
    "----------------------------------------\n",
    "Global fields:\n",
    "    'n_cases': int32 \n",
    "    'n_controls': int32 \n",
    "    'heritability': float64 \n",
    "    'saige_version': str \n",
    "    'inv_normalized': str \n",
    "    'log_pvalue': bool \n",
    "    'ranks': struct {\n",
    "        values: array<float64>, \n",
    "        ranks: array<int32>, \n",
    "        _compaction_counts: array<int32>\n",
    "    } \n",
    "    'lambda_gc': float64 \n",
    "----------------------------------------\n",
    "Row fields:\n",
    "    'CHR': str \n",
    "    'POS': int32 \n",
    "    'MarkerID': str \n",
    "    'Allele1': str \n",
    "    'Allele2': str \n",
    "    'AC_Allele2': int32 \n",
    "    'AF_Allele2': float64 \n",
    "    'MissingRate': float64 \n",
    "    'BETA': float64 \n",
    "    'SE': float64 \n",
    "    'var': float64 \n",
    "    'p.value.NA': float64 \n",
    "    'Is.SPA': bool \n",
    "    'AF_case': float64 \n",
    "    'AF_ctrl': float64 \n",
    "    'locus': locus<GRCh38> \n",
    "    'alleles': array<str> \n",
    "    'phenoname': str \n",
    "    'Pvalue': float64 \n",
    "    'Pvalue_log10': float64 \n",
    "    'rank': int64 \n",
    "    'Pvalue_expected': float64 \n",
    "    'Pvalue_expected_log10': float64 \n",
    "----------------------------------------\n",
    "Key: ['locus', 'alleles', 'phenoname']\n",
    "----------------------------------------\n",
    "All by All Resources\n",
    "If you are unsure which phecode to use for your Hail Table import, please reference the phenotype index provided by All of Us.\n",
    "\n",
    "To learn more about the All by All tables, check out the Overview of All by All Tables in the User Support Hub.\n",
    "\n",
    "For each phenotype category, there are multiple notebooks that offer basic summary statistics for every phenotype in the All by All tables! You can find them under Featured Workspaces.a"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "",
   "name": ""
  },
  "language_info": {
   "name": ""
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
